fix: ClientRpcs ignores NetworkObject's observer list [MTT-2878]

NoelStephensUnity opened the pull request:
This fixes the issue where NetworkBehaviour.__endSendClientRpc ignored the associated NetworkObject's Observers list and would send to all connected clients (including non-observers which would not have that NetworkObject instance).

## Changelog
- Fixed: ClientRpcs would always send to all connected clients by default as opposed to only sending to the NetworkObject's Observers list by default.

## Testing and Documentation
- Includes an integration test.

NoelStephensUnity committed:
fix

First pass fix for bug MTT-2878 where ClientRPCs do not take into account the NetworkObject's observers list

NoelStephensUnity committed:
update

Adding observer checks when using ClientRpcParams.Send.TargetClientIds or ClientRpcParams.Send.TargetClientIdsNativeArray to let users know that they are sending a ClientRpc to a client that is not an observer of the NetworkObject.

NoelStephensUnity committed:
test

This test verifies that only client observers of a NetworkObject receive the ClientRpc message.

NoelStephensUnity committed:
update

Created GenerateObserverErrorMessage for generating the "sending to a non-observer" error message.

NoelStephensUnity committed:
test

updated the test to check for the "sending to a non-observer" error message

NoelStephensUnity committed:
Merge branch 'develop' into fix/mtt-2878-clientrpc-ignores-observers

NoelStephensUnity committed:
Update CHANGELOG.md

NoelStephensUnity committed:
update

Added check to see if only the host was supposed to receive the rpc when it was the only observer and check to see if the host received the rpc when no clients were connected.

NoelStephensUnity committed:
style

updated comments and assert message.

becksebenius-unity commented on com.unity.netcode.gameobjects/Runtime/Core/NetworkBehaviour.cs:
```diff
@@ -228,6 +268,13 @@ internal void __endSendClientRpc(ref FastBufferWriter bufferWriter, uint rpcMeth
 #endif
         }
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
```
Curious why to add this? Does it do something important?

NoelStephensUnity replied:
The compiler will try to inline the method (if possible) which if inlined there is no impact to the stack and it is faster.
[Here is a good overview on this ]( https://ireadable.wordpress.com/2017/08/02/methodimplmethodimploptions-aggressiveinlining-2/)where it does a comparison of two identical methods one has the attribute and the other doesn't.  The one without consumes 7.34ns and the one with the attribute consumes 0.32ns.

becksebenius-unity replied:
I'm curious if you have benchmarked this particular case though, because without this attribute the compiler will still try to inline where it makes sense and conventional wisdom in c# is that we rarely know better than the compiler of when it's the best time to inline a method

NoelStephensUnity replied:
Yeah, I can see that.  I will remove that attribute.

NoelStephensUnity committed:
update

Removing the [MethodImpl(MethodImplOptions.AggressiveInlining)] attribute.

NoelStephensUnity committed:
style

removing namespace.

NoelStephensUnity committed:
Merge branch 'develop' into fix/mtt-2878-clientrpc-ignores-observers

TwoTenPvP commented on com.unity.netcode.gameobjects/Runtime/Core/NetworkBehaviour.cs:
```diff
@@ -168,6 +168,12 @@ internal void __endSendClientRpc(ref FastBufferWriter bufferWriter, uint rpcMeth
+                    // Check to make sure we are sending to only observers, if not log an error.
+                    if (!NetworkObject.Observers.Contains(targetClientId))
+                    {
+                        Debug.LogError(GenerateObserverErrorMessage(clientRpcParams, targetClientId));
```
Should use NetworkLog and should check the log level.

TwoTenPvP commented on com.unity.netcode.gameobjects/Runtime/Core/NetworkBehaviour.cs:
```diff
@@ -181,14 +187,47 @@ internal void __endSendClientRpc(ref FastBufferWriter bufferWriter, uint rpcMeth
+                    // Check to make sure we are sending to only observers, if not log an error.
+                    if (!NetworkObject.Observers.Contains(targetClientId))
+                    {
+                        Debug.LogError(GenerateObserverErrorMessage(clientRpcParams, targetClientId));
```
Same as above

TwoTenPvP commented on com.unity.netcode.gameobjects/Runtime/Core/NetworkBehaviour.cs:
```diff
@@ -181,14 +187,47 @@ internal void __endSendClientRpc(ref FastBufferWriter bufferWriter, uint rpcMeth
+                    }
+
+                    rpcWriteSize = NetworkManager.SendMessage(ref clientRpcMessage, networkDelivery, observerClientIds);
+                    observerClientIds.Dispose();
```
Perhaps "using" should be used to prevent leaks.

TwoTenPvP reviewed (changes requested):
Minor changes, overall looks neat!

TwoTenPvP committed:
Merge branch 'develop' into fix/mtt-2878-clientrpc-ignores-observers

NoelStephensUnity committed:
update

addressing Albin's catches on how we log an error as well as assuring the array is disposed.

NoelStephensUnity committed:
update

Setting the log level required to LogLevel.Error

NoelStephensUnity committed:
fix

Greater than or equal to Error...that is.  >.<

TwoTenPvP reviewed (approved):
After the changes discussed here and in Slack. 👌🏻

NoelStephensUnity committed:
Merge branch 'develop' into fix/mtt-2878-clientrpc-ignores-observers

NoelStephensUnity committed:
fix

Adjusting RpcObserverTests error message check since we changed it to a Networklog.

NoelStephensUnity committed:
Merge branch 'develop' into fix/mtt-2878-clientrpc-ignores-observers

mattwalsh-unity commented on com.unity.netcode.gameobjects/Runtime/Core/NetworkBehaviour.cs:
```diff
@@ -158,7 +158,7 @@ internal void __endSendClientRpc(ref FastBufferWriter bufferWriter, uint rpcMeth
             // to ourself. Sadly we have to figure that out from the list of clientIds :(
             bool shouldSendToHost = false;
-
+            var logLevel = NetworkManager.LogLevel;
```
Though I'm a big fan of temp variables for clarity, I'm not sure this actually does so here.

mattwalsh-unity commented on com.unity.netcode.gameobjects/Runtime/Core/NetworkBehaviour.cs:
```diff
@@ -158,7 +158,7 @@ internal void __endSendClientRpc(ref FastBufferWriter bufferWriter, uint rpcMeth
             bool shouldSendToHost = false;
-
+            var logLevel = NetworkManager.LogLevel;
             if (clientRpcParams.Send.TargetClientIds != null)
```
I understand why we cannot readily fuse the if and else loops, but it still makes me sad.

NoelStephensUnity committed:
Merge branch 'develop' into fix/mtt-2878-clientrpc-ignores-observers

mattwalsh-unity commented on com.unity.netcode.gameobjects/Runtime/Core/NetworkBehaviour.cs:
```diff
@@ -181,14 +187,56 @@ internal void __endSendClientRpc(ref FastBufferWriter bufferWriter, uint rpcMeth
+                                shouldSendToHost = true;
+                                continue;
+                            }
+                            observerClientIds[observerCount++] = observerEnumerator.Current;
```
Rather than go to the trouble of allocating and building up the client ids, why not simply here in the while loop do:

```
NetworkManager.SendMessage(ref clientRpcMessage, networkDelivery, observerEnumerator.Current)
```

NoelStephensUnity replied:
After our internal discussion, I changed my approach to avoid the temporary memory allocation.

NoelStephensUnity committed:
update

Removed var logLevel.
Re-organized the error logging check such that it will only perform the HashSet.Contains call when LogLevel.Error level or higher.

NoelStephensUnity committed:
update

removing the allocation, avoiding cost of using NetworkManager.SendMessage, and now adding each observer client's message to the outbound queue one at a time.
